File: day-2/part-1.ts
import { loadInput } from "../utils/utils";

type Move = "forward" | "up" | "down";
type Command = `${Move} ${number}`;
interface Position {
    h: number;
    d: number;
}

const performCommand = (comm: Command, { h, d }: Position) => {
    const [move, strStep] = comm.split(" ") as [Move, string];
    const step = parseInt(strStep);

    if(move === "forward") return { h: h + step, d };
    if(move === "up") return { h, d: d - step };
    if(move === "down") return { h, d: d + step };

    return { h, d };
} 

const followCommands = (commands: Command[], startingPosition: Position = { h: 0, d: 0 }) => {
    const finalPosition = commands.reduce(
        (position, command) => performCommand(command, position),
        startingPosition
    );

    return finalPosition;
};

const calcPosition = ({ h, d }: Position) => h * d;

(() => {
    const input = loadInput("input") as Command[];
    
    const finalPosition = followCommands(input);
    const result = calcPosition(finalPosition)
    
    console.log(result);
})();

File: day-2/part-1.py

def load_input(file):
    with open(file, "r") as lines:
        return [line for line in lines]

def execute_move(move, step, position):
    delta = {}

    if move == "forward": delta = { "horizontal": step }
    if move == "down": delta = { "depth": step }
    if move == "up": delta = { "depth": -step }

    for key, val in delta.items():
        position[key] += val
    
    return position

def perform_moves(data):
    position = {
        "horizontal": 0,
        "depth": 0
    }

    for line in data:
        move, step = line.split(" ")
        position = execute_move(move, int(step), position)
        
    return position

def main():
    input = load_input("input")

    final_position = perform_moves(input)
    result = final_position["horizontal"] * final_position["depth"]
    
    print(result)

if __name__ == "__main__":
    main()
    